### DISCUSSION TITLE
New SwarmBuilder usage experience

### DISCUSSION THREAD

**Discussion by gurinderu** [libp2p/rust-libp2p#4719] [General]
All current examples have a simple code defined in one function. Let's imagine I would like to make a function decomposition with the signature `fn someLogic(builder: SwarmBuilder<_, _>): SwarmBuilder<_, _>`, but it is not possible to return and pass it if the provider or phase is public only for internal modules.

**Comment by mxinden** [comment]
Instead of passing around different stages of `SwarmBuilder`, would it be possible for you to pass around the information needed (e.g. with a set of `enum`s) to then construct the  `Swarm` with the `SwarmBuilder` in one final step within a single function?

If not, can you provide us with some code snippets so that we can better understand your use-case?
